File: BACK-END/KuentecoApp/src/main/java/org/kuenteco/backend/dto/logic/category/CategoryReportDTO.java
package org.kuenteco.backend.dto.logic.category;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.kuenteco.backend.dto.logic.budget.BudgetDTO;
import org.kuenteco.backend.dto.logic.transaction.kuenteco.TransactionDetailDTO;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class CategoryReportDTO {
    private CategoryDTO category;
    private BudgetDTO budget;
    private BigDecimal totalSpent;
    private BigDecimal totalIncome;
    private BigDecimal categoryRemainingBudget;
    private Integer transactionCount;
    private List<TransactionDetailDTO> transactions;
    private LocalDateTime reportGeneratedAt;
}

File: FRONT-END/lib/dto/category/category_report_dto.dart
import '../budget/budget_dto.dart';
import '../transaction/kuenteco/transaction_detail_dto.dart';
import 'category_dto.dart';

class CategoryReportDTO {
  final CategoryDTO? category;
  final BudgetDTO? budget;
  final double? totalSpent;
  final double? totalIncome;
  final double? categoryRemainingBudget;
  final int? transactionCount;
  final List<TransactionDetailDTO>? transactions;
  final DateTime? reportGeneratedAt;

  CategoryReportDTO({
    this.category,
    this.budget,
    this.totalSpent,
    this.totalIncome,
    this.categoryRemainingBudget,
    this.transactionCount,
    this.transactions,
    this.reportGeneratedAt,
  });

  factory CategoryReportDTO.fromJson(Map<String, dynamic> json) {
    return CategoryReportDTO(
      category: json['category'] != null
          ? CategoryDTO.fromJson(json['category'])
          : null,
      budget: json['budget'] != null
          ? BudgetDTO.fromJson(json['budget'])
          : null,
      totalSpent: (json['totalSpent'] as num?)?.toDouble(),
      totalIncome: (json['totalIncome'] as num?)?.toDouble(),
      categoryRemainingBudget:
      (json['categoryRemainingBudget'] as num?)?.toDouble(),
      transactionCount: json['transactionCount'],
      transactions: json['transactions'] != null
          ? (json['transactions'] as List)
          .map((e) => TransactionDetailDTO.fromJson(e))
          .toList()
          : null,
      reportGeneratedAt: json['reportGeneratedAt'] != null
          ? DateTime.parse(json['reportGeneratedAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'category': category?.toJson(),
      'budget': budget?.toJson(),
      'totalSpent': totalSpent,
      'totalIncome': totalIncome,
      'categoryRemainingBudget': categoryRemainingBudget,
      'transactionCount': transactionCount,
      'transactions': transactions?.map((e) => e.toJson()).toList(),
      'reportGeneratedAt': reportGeneratedAt?.toIso8601String(),
    };
  }
}
